pretyping/pretyping.ml

```diff
@@ -845,29 +864,19 @@ struct
           in
           let sigma, ujval = adjust_evar_source sigma na.binder_name ujval in
           let value, typ = app_f n (j_val resj) ujval, subst1 ujval c2 in
-          let j = { uj_val = value; uj_type = typ } in
-          apply_rec env sigma (n+1) j candargs bidiargs rest
+          let resj = { uj_val = value; uj_type = typ } in
+          let resj_before_bidi = if bidi then resj_before_bidi else resj in
+          apply_rec env sigma (n+1) resj resj_before_bidi candargs bidiargs rest
         | _ ->
           let sigma, hj = pretype empty_tycon env sigma c in
           error_cant_apply_not_functional
             ?loc:(Loc.merge_opt floc argloc) !!env sigma resj [|hj|]
     in
-    let sigma, resj, bidiargs = apply_rec env sigma 0 fj candargs [] args in
-    let sigma, resj =
-      match EConstr.kind sigma resj.uj_val with
-      | App (f,args) ->
-        if Termops.is_template_polymorphic_ind !!env sigma f then
-          (* Special case for inductive type applications that must be
-             refreshed right away. *)
-          let c = mkApp (f, args) in
-          let sigma, c = Evarsolve.refresh_universes (Some true) !!env sigma c in
-          let t = Retyping.get_type_of !!env sigma c in
-          sigma, make_judge c (* use this for keeping evars: resj.uj_val *) t
-        else sigma, resj
-      | _ -> sigma, resj
-    in
-    let sigma, t = inh_conv_coerce_to_tycon ?loc ~program_mode resolve_tc env sigma resj tycon in
-    let refine_arg sigma (newarg,origarg) =
+    let sigma, resj, resj_before_bidi, bidiargs = apply_rec env sigma 0 fj fj candargs [] args in
+    let sigma, resj = refresh_template env sigma resj in
+    let sigma, resj_before_bidi = refresh_template env sigma resj_before_bidi in
```

Comment:
I think we should refresh only after having refined args, in the `otrace = Some trace` branch before reapply the coercion. Refresh is supposed to be done after all the arguments are applied, and in the None branch we would discard the result so it doesn't matter.
Also I would like to avoid double refreshing when there is no bidi (vast majority of cases).

We could also say that it's nonsensical to have bidi hints on template inductives, this could help make things simpler.